packages/autoloader/src/AutoloadGenerator.php

```diff
@@ -0,0 +1,372 @@
+	 * @param InstalledRepositoryInterface $localRepo
+	 * @param PackageInterface             $mainPackage
+	 * @param InstallationManager          $installationManager
+	 * @param string                       $targetDir
+	 * @param bool                         $scanPsr0Packages
+	 * @param string                       $suffix
+	 */
+	public function dump(
+		Config $config,
+		InstalledRepositoryInterface $localRepo,
+		PackageInterface $mainPackage,
+		InstallationManager $installationManager,
+		$targetDir,
+		$scanPsr0Packages = false,
+		$suffix = ''
+	) {
+
+		if ( $this->classMapAuthoritative ) {
+
+			// Force scanPsr0Packages when classmap is authoritative
+			$scanPsr0Packages = true;
+		}
+
+		$this->devMode = $scanPsr0Packages;
+
+		$filesystem = new Filesystem();
+		$filesystem->ensureDirectoryExists( $config->get( 'vendor-dir' ) );
+
+		$basePath   = $filesystem->normalizePath( realpath( getcwd() ) );
+		$vendorPath = $filesystem->normalizePath( realpath( $config->get( 'vendor-dir' ) ) );
+		$targetDir  = $vendorPath . '/' . $targetDir;
+		$filesystem->ensureDirectoryExists( $targetDir );
+
+		$vendorPathCode = $filesystem->findShortestPathCode( realpath( $targetDir ), $vendorPath, true );
+
+		$appBaseDirCode = $filesystem->findShortestPathCode( $vendorPath, $basePath, true );
+		$appBaseDirCode = str_replace( '__DIR__', '$vendorDir', $appBaseDirCode );
+
+		$packageMap = $this->buildPackageMap( $installationManager, $mainPackage, $localRepo->getCanonicalPackages() );
+		$autoloads  = $this->parseAutoloads( $packageMap, $mainPackage );
+
+		// add custom psr-0 autoloading if the root package has a target dir
+		$targetDirLoader = null;
+
+		$filesCode = '';
+		// TODO: Test out file autoloading
+		// $autoloads['files'] = new \RecursiveIteratorIterator(new \RecursiveArrayIterator($autoloads['files']));
+		// foreach ($autoloads['files'] as $functionFile) {
+		// don't include file if it is using PHP 5.3+ syntax
+		// https://bitbucket.org/xrstf/composer-php52/issue/4
+		// if ($this->isPHP53($functionFile)) {
+		// $filesCode .= '//       require ' . $this->getPathCode($filesystem, $basePath, $vendorPath,
+		// $functionFile) . "; // disabled because of PHP 5.3 syntax\n";
+		// } else {
+		// $filesCode .= '     require ' . $this->getPathCode($filesystem, $basePath, $vendorPath,
+		// $functionFile) . ";\n";
+		// }
+		// }
+
+		$classmapFile = <<<EOF
+<?php
+
+// autoload_classmap_package.php @generated by automattic/autoload
+
+\$vendorDir = $vendorPathCode;
+\$baseDir = $appBaseDirCode;
+
+return array(
+
+EOF;
+
+		$blacklist = null;
+		if ( ! empty( $autoloads['exclude-from-classmap'] ) ) {
+			$blacklist = '{(' . implode( '|', $autoloads['exclude-from-classmap'] ) . ')}';
+		}
+
+		// flatten array
+		$classMap = array();
+		if ( $scanPsr0Packages ) {
+			$namespacesToScan = array();
+
+			// Scan the PSR-0/4 directories for class files, and add them to the class map
+			foreach ( array( 'psr-0', 'psr-4' ) as $psrType ) {
```

Comment:
Are we planning to use psr-0?

Reply:
We are not, just use psr-4 please.

Reply:
I stripped out a lot of the different ways that composer packages can be defined to autoload and only supporting packages that load via 'psr-4' for now.